### PR TITLE
#18 #14 Ft update product

### PR DESCRIPTION
**Pull request by Shumba250**
#### What does this PR do?

- a seller should update a specific item in a collection
- only the owner of the product is authorized to update a specific item
- a seller can update an existing product
- all inputs are validated 
- appropriate error messages are displayed upon the failure or success of the request
#### Description of Task to be completed

- a seller is authorized to update a product 
- all inputs are validated
- The appropriate 4XX response message is sent to the requesting client describing the validation error
#### How should this be manually tested?

- clone the repo
- navigate to the ft-update-product branch
- install packages by typing npm install
- include Cloudinary variables in .env cloud_name = "" ,  api_key = "", api_secret = ""
- start the server by typing npm run dev
- open your browser to use swagger doc by navigating to localhost/5050/docs
- create vendor a role and permissions
- register a vendor or if you have one created, log the vendor in
- after completing the 2FA authentication, the vendor with be logged in automatically
- on swagger doc, use create product endpoint and only a minimum of 4 images are allowed and create a product
- after creating a product, use update product endpoint and use the product id to update a product

### CODE REVIEW

**Review by sostene12 — APPROVED** [review]

File: src/controllers/userController.js [hunk 1]
```diff
@@ -86,6 +85,7 @@ class UserController {
       return res.status(400).json({ status: "error", error: error.message });
```
Comment (wakaflorien) [inline src/controllers/userController.js]: - for consitency and uniformity purposes use something like this on responses instead:

`res.status(400).json({ status: "fail", error: error.message });`

File: src/controllers/userController.js [hunk 2]
```diff
@@ -109,6 +109,7 @@ class UserController {
       return res.status(500).json({ status: "error", error: error.message });
```
Comment (wakaflorien) [inline src/controllers/userController.js]: - same here

File: src/middlewares/productExists.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import { Product } from "../database/models";
+
+export async function IsProductExist(req, res, next) {
+  try {
+    const productId = req.params.id || req.body.product_id;
+    const product = await Product.findOne({ where: { id: productId } });
+    if (!product || product == null) {
+      return res
+        .status(404)
+        .json({ status: 404, message: "product does not exist" });
```
Comment (wakaflorien) [inline src/middlewares/productExists.js]: - avoid redudant status codes do something like thsi instead:

`res
        .status(404)
        .json({ status: "fali", message: "product does not exist" });`

File: src/middlewares/productExists.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    req.product = product;
+    next();
+  } catch (error) {
+    return res.status(500).json({ status: 500, error: "server error" });
```
Comment (wakaflorien) [inline src/middlewares/productExists.js]: ### same here

File: src/middlewares/productExists.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+export async function productExistAlready(req, res, next) {
+  try {
+    const product = await Product.findOne({
+      where: { productName: req.body.productName }
+    });
+    if (product) {
+      return res
+        .status(409)
```
Comment (wakaflorien) [inline src/middlewares/productExists.js:27]: - ### do the same here

File: src/middlewares/productExists.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        .json({ status: 409, message: "product already exists" });
+    }
+    req.product = product;
+    next();
+  } catch (error) {
+    return res.status(500).json({ status: 500, error: error.message });
```
Comment (wakaflorien) [inline src/middlewares/productExists.js]: - ### again

**Review by wakaflorien — CHANGES_REQUESTED** [review]

**Review by wakaflorien — APPROVED** [review]

**Review by oliviertech212 — APPROVED** [review]

**Review by chrill-christella — APPROVED** [review]

**Pull request merged**